### PR TITLE
Disintegrating Microsoft.ML.Api dispersing its content.

### LINKED ISSUE #1707: Merge API assembly into Data
Go over `Microsoft.ML.Api` in the following fashion:

- For all code that is currently used by our code or tests, move it to `Microsoft.ML.Data`.
  - This, I assume, will include all schema propagation, `PredictionEngine` (but not Batch), and `CustomMappingEstimator` and probably the `StatefulFilter` for the time series.
- For all code that is needed by TLC, move back to TLC.
  - I'm not aware of any such code at the moment.
- For code that is not used by either, remove.

After this, remove `Microsoft.ML.Api`

### PR DESCRIPTION
**Pull request by sfilipi**
Fixes #1707  by moving most of the classes on it to Microsoft.ML.Data or Microsoft.ML.Core. 

Let me know if there is a better place for particular files.

### CODE REVIEW

File: src/Microsoft.ML.Api/CodeGenerationUtils.cs [hunk 1]
```diff
@@ -1,142 +0,0 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
-// See the LICENSE file in the project root for more information.
-
-using System.Collections.Generic;
```
Comment (sfilipi) [inline src/Microsoft.ML.Api/CodeGenerationUtils.cs]: >using System.Collections.Gener [](start = 0, length = 30)

@yaeldekel is it a good time to delete this?

**Comment by sfilipi** [comment]
@TomFinley, @yaeldekel @Ivanidzo4ka @najeeb-kazmi @wschin It looks like a lot of files, but it is actually an easy review. Just look at the files moved, if they are in the right place, and the namespace. 

The other files (not moved) are just using fixes.

**Comment by sfilipi** [comment]
Revert the changes to libmf submodule.

**Comment by Ivanidzo4ka** [comment]
>     <TargetFramework>netstandard2.0</TargetFramework>

You deleted this project from solution. Why we keep it here?

---
Refers to: src/Microsoft.ML.Api/Microsoft.ML.Api.csproj:4 in 6337f66. [](commit_id = 6337f663d841b5709a5567c21809ddbf51327f22, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>     public sealed class Predictor

I would nuke this file and replace everything with IPredictor.

---
Refers to: src/Microsoft.ML.Data/Prediction/Predictor.cs:13 in 6337f66. [](commit_id = 6337f663d841b5709a5567c21809ddbf51327f22, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>     public sealed class BatchPredictionEngine<TSrc, TDst>

internal + bestfriend?

---
Refers to: src/Microsoft.ML.Data/Prediction/PredictionEngine.cs:22 in 6337f66. [](commit_id = 6337f663d841b5709a5567c21809ddbf51327f22, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>     public sealed class PredictionEngine<TSrc, TDst> : PredictionEngineBase<TSrc, TDst>

internal + bestfriend

---
Refers to: src/Microsoft.ML.Data/Prediction/PredictionEngine.cs:125 in 6337f66. [](commit_id = 6337f663d841b5709a5567c21809ddbf51327f22, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>     public abstract class PredictionEngineBase<TSrc, TDst>

internal + bestfriend

---
Refers to: src/Microsoft.ML.Data/Prediction/PredictionEngine.cs:160 in 6337f66. [](commit_id = 6337f663d841b5709a5567c21809ddbf51327f22, deletion_comment = False)

**Comment by Ivanidzo4ka** [comment]
>     public static class ComponentCreation

I think that whole thing is our old API, and shouldn't be exposed, so can we make it internal?

---
Refers to: src/Microsoft.ML.Data/Utilities/ComponentCreation.cs:20 in 6337f66. [](commit_id = 6337f663d841b5709a5567c21809ddbf51327f22, deletion_comment = False)

File: src/Microsoft.ML.SamplesUtils/Microsoft.ML.SamplesUtils.csproj [hunk 2]
```diff
@@ -4,9 +4,9 @@
     <TargetFramework>netstandard2.0</TargetFramework>
     <IncludeInPackage>Microsoft.ML</IncludeInPackage>
   </PropertyGroup>
-
+
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.SamplesUtils/Microsoft.ML.SamplesUtils.csproj]: *gasp

File: src/Microsoft.ML.Legacy/AssemblyRegistration.cs [hunk 3]
```diff
@@ -55,7 +53,7 @@ private static bool LoadStandardAssemblies()
             _ = Assembly.Load(new AssemblyName()
             {
                 Name = "Microsoft.ML.StandardLearners",
-                Version = apiAssemblyName.Version, //assume the same version as ML.Api
+                Version = dataAssemblyName.Version, //assume the same version as ML.Api
```
Comment (TomFinley) [inline src/Microsoft.ML.Legacy/AssemblyRegistration.cs]: No biggie just an out of date comment.

**Comment by eerhardt** [comment]
>     <TargetFramework>netstandard2.0</TargetFramework>

Yes, let's delete this file if possible.

---
In reply to: [445345496](https://github.com/dotnet/machinelearning/pull/1848#issuecomment-445345496) [](ancestors = 445345496)

---
Refers to: src/Microsoft.ML.Api/Microsoft.ML.Api.csproj:4 in 6337f66. [](commit_id = 6337f663d841b5709a5567c21809ddbf51327f22, deletion_comment = False)

File: src/Microsoft.ML.Legacy/AssemblyRegistration.cs [hunk 4]
```diff
@@ -38,10 +37,9 @@ public static void RegisterAssemblies(IHostEnvironment environment)
         /// </summary>
         private static bool LoadStandardAssemblies()
         {
-            Assembly apiAssembly = typeof(LambdaTransform).Assembly; // ML.Api
-            AssemblyName apiAssemblyName = apiAssembly.GetName();
+            Assembly dataAssembly = typeof(TextLoader).Assembly; // ML.Data
+            AssemblyName dataAssemblyName = dataAssembly.GetName();
 
-            _ = typeof(TextLoader).Assembly; // ML.Data
             //_ = typeof(EnsemblePredictor).Assembly); // ML.Ensemble BUG https://github.com/dotnet/machinelearning/issues/1078 Ensemble isn't in a NuGet package
```
Comment (eerhardt) [inline src/Microsoft.ML.Legacy/AssemblyRegistration.cs]: While you are in here - can you uncomment this line and remove the "BUG" comment?

**Comment by eerhardt** [comment]
The `libmf` change seems unrelated. Let's not update `libmf` if possible.

**Comment by sfilipi** [comment]
>     public abstract class PredictionEngineBase<TSrc, TDst>

turns out i have to keep this public. It is the base class for the time series Prediction engine, which is public. 

---
In reply to: [445360511](https://github.com/dotnet/machinelearning/pull/1848#issuecomment-445360511) [](ancestors = 445360511)

---
Refers to: src/Microsoft.ML.Api/PredictionEngine.cs:160 in 6337f66. [](commit_id = 6337f663d841b5709a5567c21809ddbf51327f22, deletion_comment = False)

File: src/Microsoft.ML.Api/CodeGenerationUtils.cs [hunk 5]
```diff
@@ -1,142 +0,0 @@
...
-using Microsoft.ML.Runtime.Data;
-using Microsoft.CSharp;
-using System.IO;
-
-namespace Microsoft.ML.Runtime.Api
-{
-    /// <summary>
-    /// Utility methods for code generation.
-    /// </summary>
-    internal static class CodeGenerationUtils
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Api/CodeGenerationUtils.cs]: >CodeGenerationUtils [](start = 26, length = 19)

you need to do something with

 [Fact]
        public void CommandCodeGen()
        {
            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
                return;

Comment (eerhardt) [reply src/Microsoft.ML.Api/CodeGenerationUtils.cs]: Delete it.

Comment (sfilipi) [reply src/Microsoft.ML.Api/CodeGenerationUtils.cs]: yes, saw that failing. thanks. 

---
In reply to: [240723443](https://github.com/dotnet/machinelearning/pull/1848#discussion_r240723443) [](ancestors = 240723443)

**Comment by sfilipi** [comment]
> // Licensed to the .NET Foundation under one or more agreements.

thanks for the catch, @eerhardt

---
In reply to: [446014349](https://github.com/dotnet/machinelearning/pull/1848#issuecomment-446014349) [](ancestors = 446014349)

---
Refers to: docs/samples/Microsoft.ML.Samples/Dynamic/TensorFlowTransform.cs:1 in 731c6f1. [](commit_id = 731c6f14abf06910c5e5412d8f3dd8d2ebc0a804, deletion_comment = False)

**Pull request merged**